### PR TITLE
[WIP] Nanochatgpt edit - rebased

### PR DESCRIPTION
**Pull request by tcbegley**
## Description

This contains the changes of #29 but rebased onto nanoChatGPT

### CODE REVIEW

File: examples/nanoChatGPT/train_rlhf.py [hunk 1]
```diff
@@ -64,32 +66,38 @@ def forward(self, tensordict):
...
-        loss_val = sum(
-            value for key, value in loss_vals.items() if key.startswith("loss")
+            td = env.rollout(ep_length, policy=actor, return_contiguous=True).cpu()
+        for epoch in range(num_epochs):
+            tdd = adv_fn(td.to(device))
+            rb.extend(tdd.view(-1))
+            del tdd
+            for batch in rb:
+
+                # TODO: add replay buffer?
```
Comment (apbard) [inline examples/nanoChatGPT/train_rlhf.py:88]: remove this

File: examples/nanoChatGPT/train_rlhf.py [hunk 2]
```diff
@@ -64,32 +66,38 @@ def forward(self, tensordict):
...
+                loss_vals = loss_fn(batch.to(device))
+
+                loss_val = sum(
+                    value for key, value in loss_vals.items() if key.startswith("loss")
+                )
+                loss_val.backward()
+                optimizer.step()
+                optimizer.zero_grad()
+        print(
+            f"Iteration {i}: {loss_val=}, reward={td.get(('next', 'reward')).mean(): 4.4f}"
```
Comment (apbard) [inline examples/nanoChatGPT/train_rlhf.py:99]:
```suggestion
            f"Iteration {i}: {loss_val=}, reward={-td.get(('next', 'reward')).mean(): 4.4f}"
```

File: examples/nanoChatGPT/train_rlhf.py [hunk 3]
```diff
@@ -34,24 +36,24 @@ def __init__(self, critic):
...
+
     vmap_critic = VmapCritic(critic)
 
     adv_fn = GAE(value_network=vmap_critic, gamma=0.99, lmbda=0.95, average_gae=True)
-    
+
     # FIXME: why not using the scheduler?
     # Loss
-    loss_fn = ClipPPOLoss(actor, critic, gamma=0.99)
+    loss_fn = ClipPPOLoss(actor, critic_head)
```
Comment (apbard) [inline examples/nanoChatGPT/train_rlhf.py:56]: is it right we removed gamma?

File: examples/nanoChatGPT/train_rlhf.py [hunk 4]
```diff
@@ -64,32 +66,38 @@ def forward(self, tensordict):
...
-    #     td["sample_log_prob"] = td["sample_log_prob"].detach()
-    #     return td
-
-    for i in range(config["max_iters"]):
+    ep_length = config["episode_length"]
+    max_iters = config["max_iters"]
+    num_epochs = config["num_epochs"]
+    device = config['device']
+    rb = TensorDictReplayBuffer(
+        storage=LazyTensorStorage(ep_length * config["batch_size"]),
```
Comment (apbard) [inline examples/nanoChatGPT/train_rlhf.py:74]: @vmoens can you explain these buffer settings?

**Review by apbard — APPROVED** [review]
a few minor things. LGTM

File: examples/nanoChatGPT/train_rlhf.py (hunk 3 above)
Comment (vmoens) [reply examples/nanoChatGPT/train_rlhf.py:56]: yes it's only used to calculate the advantage which we do in the advantage module
If you're using the latest torchrl you should get a warning that tells you not to use gamma in the constructor

File: examples/nanoChatGPT/train_rlhf.py (hunk 4 above)
Comment (vmoens) [reply examples/nanoChatGPT/train_rlhf.py:74]: sure
We want a storage that is as big as the number of steps collected, such that when we get a rollout and put it in the buffer it erases the previous rollout completely.
We use a sampler without replacement such that we can iterate over the buffer and collect samples without replacement (consuming the buffer, so to say).
we set the batch size of the buffer to some value that determines how many samples we feed to the loss.

File: examples/nanoChatGPT/train_rlhf.py (hunk 2 above)
Comment (vmoens) [reply examples/nanoChatGPT/train_rlhf.py]: why?
